services/web-server/src/utils/verifyJwt.js

```diff
@@ -0,0 +1,39 @@
+const assert = require('assert');
+const jwt = require('jsonwebtoken');
+const jwks = require('jwks-rsa');
+
+module.exports = async ({ token, domain }) => new Promise((resolve, reject) => {
+  assert(token, 'No authorization token was found');
+  assert(domain, 'A domain is required to verify the jwt');
+
+  const client = jwks({
+    cache: true,
+    rateLimit: true,
+    jwksRequestsPerMinute: 5,
+    jwksUri: `https://${domain}/.well-known/jwks.json`,
+  });
+  const getKey = (header, callback) => {
+    client.getSigningKey(header.kid, (err, key) => {
+      const signingKey = key.publicKey || key.rsaPublicKey;
+
+      callback(null, signingKey);
+    });
+  };
+
+  jwt.verify(
+    token,
+    getKey,
+    {
+      // and expect a token issued by auth0
+      issuer: `https://${domain}/`,
+      algorithms: ['RS256'],
```

Comment:
I think we need to [check the audience, too](https://github.com/taskcluster/taskcluster/blob/master/services/login/src/handlers/mozilla-auth0.js#L32)?

Reply:
I thought so too but I was getting a weird `aud` value. See https://github.com/mozilla-iam/cis/issues/423.

Reply:
Yeah, the audience and scope are part of the way we support third-party logins.  Basically when the frontend (be that tc-ui or treeherder or releng-services) requests an auth0 login, it indicates an additional audience and scope that correspond to an auth0 "API" object which we asked the IAM team to create.  When Auth0 sees this, it presents a consent dialog saying "do you want to grant treeherder access to Taskcluster?"  The result is an access token that has two audiences (treeherder and Taskcluster), and both of those services can rely on it to authenticate the user.  So this is how we support third-party logins.

It's extra work to get the IAM team to set this up, and since we don't want to support this method of third-party logins, it's probably not worth it.

However, we do still need to verify that the access_token was meant for us - that is, that we are the intended audience!  That is done by checking that the audience is equal to the auth0 client id.  Without this, I could steal your TC credentials by:
 * setting up some toy Mozilla site (maybe another Bugzilla dashboard!)
 * configuring auth0 for that site
 * getting you to login
 * taking the resulting access_token and calling tc-web-server's `getCredentials` query with it